File: protocol/messages.ts
/**
 * Shared WebSocket message protocol between frontend and backend.
 * This file is the source of truth — the Rust backend mirrors these types via serde.
 */

// ─── Common Types ────────────────────────────────────────────────────────────

export type Color = "white" | "black";
export type BoardId = "a" | "b";
export type Square = string; // e.g. "e4", "d7"
export type Role = "pawn" | "knight" | "bishop" | "rook" | "queen" | "king";

export interface Reserve {
  pawn: number;
  knight: number;
  bishop: number;
  rook: number;
  queen: number;
}

export interface GameResult {
  /** The color that won, or null for a draw. */
  winner: Color | null;
  termination: "checkmate" | "timeout" | "resignation" | "abandoned" | "draw";
}

// ─── Client → Server Messages ────────────────────────────────────────────────

export type ClientMessage =
  | { type: "move"; from: Square; to: Square; promotion?: Role }
  | { type: "drop"; role: Role; to: Square }
  | { type: "ping" };

// ─── Server → Client Messages ────────────────────────────────────────────────

export type ServerMessage =
  | { type: "game_start"; color: Color; fen: string }
  | {
      type: "move";
      color: Color;
      from: Square;
      to: Square;
      promotion?: Role;
      fen: string;
    }
  | { type: "drop"; color: Color; role: Role; to: Square; fen: string }
  | { type: "reserve_update"; white: Reserve; black: Reserve }
  | { type: "game_over"; result: GameResult }
  | { type: "error"; message: string }
  | { type: "pong" };

File: backend/src/ws/messages.rs
use serde::{Deserialize, Serialize};

use crate::game::types::{Color, GameResult, Reserve, Role, Square};

/// Messages sent from client to server.
#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    /// Make a normal move (piece from one square to another).
    Move {
        from: Square,
        to: Square,
        #[serde(default)]
        promotion: Option<Role>,
    },
    /// Drop a piece from reserve onto the board.
    Drop { role: Role, to: Square },
    /// Ping for keepalive.
    Ping,
}

/// Messages sent from server to client.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    /// Game has started, here's the initial state.
    GameStart { color: Color, fen: String },
    /// A move was made on the board.
    Move {
        color: Color,
        from: Square,
        to: Square,
        #[serde(skip_serializing_if = "Option::is_none")]
        promotion: Option<Role>,
        fen: String,
    },
    /// A piece was dropped from reserve.
    Drop {
        color: Color,
        role: Role,
        to: Square,
        fen: String,
    },
    /// Reserve state updated (after a capture on the partner's board).
    ReserveUpdate { white: Reserve, black: Reserve },
    /// Game is over.
    GameOver { result: GameResult },
    /// An error occurred processing the client's message.
    Error { message: String },
    /// Pong response to client ping.
    Pong,
}
